### PR TITLE
Redis 5.0 support + better blocking operations

### PR DESCRIPTION
**Pull request by ghik**
* Blocking commands (e.g. `BLPOP`) can now be executed by `RedisNodeClient` and `RedisClusterClient` - previously they could only be executed on `RedisConnectionClient`.
  In order to support that, every `RedisNodeClient` now maintains a [separate connection pool](https://github.com/AVSystem/scala-commons/blob/redis5/commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala#L48) dedicated for blocking commands. Each blocking command reserves the connection until it's finished so that multiple concurrent blocking commands can be executed without waiting on each other to finish. Connections in the blocking pool are allocated as needed and reclaimed after a configurable period of being idle. `NodeConfig` is equipped with new options to control blocking connection pool.
* New sorted set commands: `ZPOPMIN`, `ZPOPMAX`, `BZPOPMIN`, `BZPOPMAX`
* New server commands: `CLIENT ID`, `CLIENT UNBLOCK`, `REPLICAOF` (political correctness)
* `HashKey` type member of `RedisApiSubset` renamed to `Field` so that it can serve both as the type of hash fields and stream entry fields.
* [Streams API](https://github.com/AVSystem/scala-commons/blob/redis5/commons-redis/src/main/scala/com/avsystem/commons/redis/commands/streams.scala) and a [test suite](https://github.com/AVSystem/scala-commons/blob/redis5/commons-redis/src/test/scala/com/avsystem/commons/redis/commands/StreamsApiSuite.scala) which is also a good demonstration of the API.
* Handling cluster `TRYAGAIN` errors by backing out and retrying after a delay.
* New, composable `RetryStrategy` - used for reconnecting, redirection handling and `TRYAGAIN` handling

### CODE REVIEW

File: commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala [hunk 1]
```diff
@@ -10,6 +10,7 @@ object OptArg {
     * This implicit conversion allows you to pass unwrapped values where `OptArg` is required.
     */
   implicit def argToOptArg[A](value: A): OptArg[A] = OptArg(value)
+  implicit def intToOptArgLong(int: Int): OptArg[Long] = OptArg(int)
```
Comment (ddworak) [inline commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala]: You don't seem to use this one. If you want to start adding these, there's a lot more "implicit" conversions to cover even with just numbers.

File: commons-redis/src/main/scala/com/avsystem/commons/redis/RedisApi.scala [hunk 2]
```diff
@@ -228,17 +228,17 @@ object RedisApi {
   }
 }
 
-abstract class AbstractRedisApi[Self[K0, H0, V0] <: AbstractRedisApi[Self, K0, H0, V0], K, H, V](implicit
-  val keyCodec: RedisDataCodec[K], val hashKeyCodec: RedisDataCodec[H], val valueCodec: RedisDataCodec[V])
+abstract class AbstractRedisApi[Self[K0, F0, V0] <: AbstractRedisApi[Self, K0, F0, V0], K, F, V](implicit
+  val keyCodec: RedisDataCodec[K], val fieldCodec: RedisDataCodec[F], val valueCodec: RedisDataCodec[V])
```
Comment (ddworak) [inline commons-redis/src/main/scala/com/avsystem/commons/redis/RedisApi.scala]: These don't need to become public.

File: commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala [hunk 3]
```diff
@@ -78,7 +109,8 @@ final class RedisNodeClient(
   private[redis] def nodeRemoved(): Unit = {
     val cause = new NodeRemovedException(address)
     failure = cause.opt
-    connections.foreach(_ ! Close(cause))
+    connections.foreach(_ ! RedisConnectionActor.Close(cause, stop = true))
```
Comment (ddworak) [inline commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala]: Both `Close` classes seem to always be used with `stop = true`

File: commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import akka.actor.{Actor, ActorRef, Props}
+import com.avsystem.commons.redis.NodeAddress
+import com.avsystem.commons.redis.actor.ConnectionPoolActor._
+import com.avsystem.commons.redis.config.{ConnectionConfig, NodeConfig}
+import com.avsystem.commons.redis.exception.RedisException
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.annotation.tailrec
+
+class ConnectionPoolActor(address: NodeAddress, config: NodeConfig, queue: ConcurrentLinkedDeque[QueuedConn])
```
Comment (ddworak) [inline commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala]: Are there multiple actors using the queue? From the actor's perspective it does not look like it has to be concurrent.

File: commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala [hunk 5]
```diff
@@ -41,9 +42,23 @@ final class RedisNodeClient(
...
-  private val connections = (0 until config.poolSize).iterator.map(createConnection).toArray
+  private val connections = (0 until config.poolSize).iterator.map(newConnection).toArray
   private val index = new AtomicLong(0)
 
+  private val blockingConnectionQueue = new ConcurrentLinkedDeque[QueuedConn]
+  private val blockingConnectionPool =
+    system.actorOf(Props(new ConnectionPoolActor(address, config, blockingConnectionQueue)))
+
+  private def newBlockingConnection(): Future[ActorRef] = {
+    implicit val timeout: Timeout = Timeout(1.second)
```
Comment (ddworak) [inline commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala]: Maybe just inline and pass explicitly?

File: commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            queue.addLast(last)
+          }
+        case null =>
+      }
+    }
+    loop(dequeue = false)
+  }
+}
+object ConnectionPoolActor {
+  case class QueuedConn(conn: ActorRef, enqueuedAt: Long)
```
Comment (ddworak) [inline commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala]: final?

File: commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala (hunk 1 above)
Comment (ghik) [reply commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala]: I'm using it in tests. I was hesitant to add this but it became too annoying. Passing `Int`s where `Long` is expected is very common and this promotion isn't caught by `argToOptArg`.

File: commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala (hunk 3 above)
Comment (ghik) [reply commons-redis/src/main/scala/com/avsystem/commons/redis/RedisNodeClient.scala]: I'd like to leave a possibility to close connection without stopping the actor, even though it's not used now.

File: commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala (hunk 4 above)
Comment (ghik) [reply commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala]: Queue is used directly in `RedisNodeClient` which is outside actor context.

Comment (ddworak) [reply commons-redis/src/main/scala/com/avsystem/commons/redis/actor/ConnectionPoolActor.scala]: I know that. I just believe that the actor does not have to enforce this requirement in *his* types, he only cares that it's a queue.

File: commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala (hunk 1 above)
Comment (ghik) [reply commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala]: I added `Int => OptArg[Double]` too and IMHO it should be enough. This covers most common and safe numeric promotions.

Comment (ddworak) [reply commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala]: Long -> Double could be an option as well, but I generally agree.

Comment (ghik) [reply commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala]: The fact that scala has `Long => Double` is wrong and don't want to proliferate it.

Comment (ghik) [reply commons-core/src/main/scala/com/avsystem/commons/misc/OptArg.scala]: OMG there's `Long => Float` :man_facepalming:

**Pull request merged**